src/compiler/checker.ts

```diff
@@ -16991,31 +16986,43 @@ namespace ts {
             return false;
         }
 
-        // Return true if type is of the given kind. A union type is of a given kind if all constituent types
-        // are of the given kind. An intersection type is of a given kind if at least one constituent type is
-        // of the given kind.
-        function isTypeOfKind(type: Type, kind: TypeFlags): boolean {
-            if (type.flags & kind) {
+        function isTypeOfKind(source: Type, kind: TypeFlags, excludeAny?: boolean) {
+            if (source.flags & kind) {
                 return true;
             }
-            if (type.flags & TypeFlags.Union) {
-                const types = (<UnionOrIntersectionType>type).types;
-                for (const t of types) {
-                    if (!isTypeOfKind(t, kind)) {
-                        return false;
-                    }
-                }
-                return true;
+            if (excludeAny && source.flags & (TypeFlags.Any | TypeFlags.Void | TypeFlags.Undefined | TypeFlags.Null)) {
+                // TODO: The callers who want this should really handle these cases FIRST
+                return false;
             }
-            if (type.flags & TypeFlags.Intersection) {
-                const types = (<UnionOrIntersectionType>type).types;
-                for (const t of types) {
-                    if (isTypeOfKind(t, kind)) {
-                        return true;
-                    }
-                }
+            const targets = [];

```

Comment:
```
var p1: number | string
var o = {
  [p1]() { }
  ~~~~ A computed property name must be of type 'string', 'number', ...
}
```

But I think I can fix that by adding a fallback to isTypeAssignableTo the union type in just the case of computed properties.